staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go

```diff
@@ -168,15 +187,47 @@ func (conn *Conn) SetIdleTimeout(duration time.Duration) {
 	conn.timeout = duration
 }
 
+// SetWriteDeadline sets a timeout on writing to the websocket connection. The
+// passed "duration" identifies how far into the future the write must complete
+// by before the timeout fires.
+func (conn *Conn) SetWriteDeadline(duration time.Duration) {
+	conn.ws.SetWriteDeadline(time.Now().Add(duration)) //nolint:errcheck
+}
+
 // Open the connection and create channels for reading and writing. It returns
 // the selected subprotocol, a slice of channels and an error.
 func (conn *Conn) Open(w http.ResponseWriter, req *http.Request) (string, []io.ReadWriteCloser, error) {
+	// serverCompleted is channel indicating server finished before it became ready.
+	serverCompleted := make(chan struct{})
+	// Ensure panic in spawned goroutine is propagated into the parent goroutine.
+	panicChan := make(chan any, 1)
 	go func() {
-		defer runtime.HandleCrash()
-		defer conn.Close()
+		// If websocket server returns, propagate panic if necessary.
+		defer func() {
+			if p := recover(); p != nil {
+				panicChan <- p
+			}
+			close(serverCompleted)
```

Comment:
this should only happen if we didn't panic

maybe rename to serveHTTPCompleted

```go
			if p := recover(); p != nil {
				panicChan <- p
			} else {
			  close(serverCompleted)
			}
```

Reply:
Added `if/else` so channel is only closed if we don't panic. Updated channel name to `serveHTTPComplete`.